source/simulator/core.cc

```diff
@@ -2102,11 +2108,18 @@ namespace aspect
                       case Parameters<dim>::AdvectionFieldMethod::fem_field:
                         assemble_advection_system (adv_field);
 
-                        if (iteration == 0)
+                        if (nonlinear_iteration == 0)
                           initial_composition_residual[c] = system_rhs.block(introspection.block_indices.compositional_fields[c]).l2_norm();
 
                         composition_residual[c]
```

Comment:
you could change this into `const double composition_residual = ...`, and then use this double in line 2119.